lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,409 @@
+        (Twine(info.prefix) + "_" + Twine(idx)).toVector(prefixSlot);
+      }
+      prefix = prefixSlot;
+    }
+    LLVM_DEBUG({
+      llvm::dbgs() << "\nMoving ";
+      if (!prefix.empty())
+        llvm::dbgs() << "`" << prefix << "` ";
+      llvm::dbgs() << inst << "\n";
+    });
+
+    // Add additional ports to the parent module as a replacement for the
+    // instance port signals once the instance is extracted.
+    unsigned numParentPorts = parent.getNumPorts();
+    unsigned numInstPorts = inst.getNumResults();
+
+    for (unsigned portIdx = 0; portIdx < numInstPorts; ++portIdx) {
+      // Assemble the new port name as "<prefix>_<name>", where the prefix is
+      // provided by the extraction annotation.
+      auto name = inst.getPortNameStr(portIdx);
+      auto nameAttr = StringAttr::get(
+          &getContext(),
+          prefix.empty() ? Twine(name) : Twine(prefix) + "_" + name);
+
+      PortInfo newPort{nameAttr,
+                       inst.getResult(portIdx).getType().cast<FIRRTLType>(),
+                       direction::flip(inst.getPortDirection(portIdx))};
+      newPort.loc = inst.getResult(portIdx).getLoc();
+      newPorts.push_back({numParentPorts, newPort});
+      LLVM_DEBUG(llvm::dbgs()
+                 << "- Adding port " << newPort.direction << " "
+                 << newPort.name.getValue() << ": " << newPort.type << "\n");
+    }
+    parent.insertPorts(newPorts);
+    anythingChanged = true;
+
+    // Replace all uses of the existing instance ports with the newly-created
+    // module ports.
+    for (unsigned portIdx = 0; portIdx < numInstPorts; ++portIdx) {
+      inst.getResult(portIdx).replaceAllUsesWith(
+          parent.getArgument(numParentPorts + portIdx));
+    }
+    assert(inst.use_empty() && "instance ports should have been detached");
+
+    // Move the original instance one level up such that it is right next to
+    // the instances of the parent module, and wire the instance ports up to
+    // the newly added parent module ports.
+    for (auto *instRecord :
+         instanceGraph->lookup(cast<hw::HWModuleLike>(*parent))->uses()) {
+      auto oldParentInst = cast<InstanceOp>(*instRecord->getInstance());
+      LLVM_DEBUG(llvm::dbgs() << "- Updating " << oldParentInst << "\n");
+      auto newParentInst = oldParentInst.cloneAndInsertPorts(newPorts);
+
+      // Migrate connections to existing ports.
+      for (unsigned portIdx = 0; portIdx < numParentPorts; ++portIdx)
+        oldParentInst.getResult(portIdx).replaceAllUsesWith(
+            newParentInst.getResult(portIdx));
+
+      // Add the moved instance and hook it up to the added ports.
+      ImplicitLocOpBuilder builder(inst.getLoc(), newParentInst);
+      auto newInst = inst.cloneAndInsertPorts({});
+      newInst->remove();
+      builder.setInsertionPointAfter(newParentInst);
+      builder.insert(newInst);
+      for (unsigned portIdx = 0; portIdx < numInstPorts; ++portIdx) {
```

Comment:
Consider: `for (auto portAndId : llvm::enumerate(newInst.getResuilts())) { /* ... */ }`.